### DISCUSSION TITLE
.NET 8 MAUI Push Notifications (Android) using Azure Notification Hub FCM V1 protocol

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/maui#26037] [Q&A]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/NET-8-MAUI-Push-Notifications-Android/10672198)._

---
I have recently implemented Push Notifications using Azure Notification Hubs and Firebase messaging for **.NET 8 MAUI**.

And managed to successfully get this working using Legacy GCM/FCM protocol in .NET 8 MAUI (Android).

But as we now have to migrate this over to use **FCM V1** due to Legacy FCM being out of support from **June 2024 (!) **. I have started making the migration changes to my .NET MAUI project and have run into some issues.

**Changes applied:**

*   Enabled Firebase Cloud Messaging API in Google Cloud Console  
    [Firebase.PNG](https://aka.ms/dc/file?name=B9cdc854bdd6c47eab41620057971449b638526865749965500_Firebase.PNG&tid=9cdc854bdd6c47eab41620057971449b638526865749965500)
    
*   Azure Portal to configure the settings for Google (FCM v1) in my Azure Notification Hub resource
    
*   Registering my device to Azure Notification Hub using FCM v1 protocol in my .NET MAUI project  
    [DeviceInstall.PNG](https://aka.ms/dc/file?name=B38ec574b878b44b6bac1847abf6893f6638526872066276367_DeviceInstall.PNG&tid=38ec574b878b44b6bac1847abf6893f6638526872066276367)
    
**Issue**  
[FirebaseMessagingService.PNG](https://aka.ms/dc/file?name=Bc53567403e0e437c9d1adb4ad6818214638526863890199234_FirebaseMessagingService.PNG&tid=c53567403e0e437c9d1adb4ad6818214638526863890199234)

Since applying the changes, I can no longer get the overridden method OnMessageReceived(RemoteMessage p0) to trigger anymore, implemented from Firebase.Messaging.FirebaseMessagingService.

I can see the device is registered using FCM v1 protocol, if I attempt to retrieve the installation using Microsoft.Azure.NotificationHubs.NotificationHubClient after running the code shown in the DeviceInstall.PNG image.

Wondering if there are more defined/documented steps to migrate this across or if anyone can help me get this back up and working using FCM v1?

Any help would be much appreciated! Thanks!

---
#### Feedback Bot on 5/30/2024, 06:39 PM: 

(private comment, text removed)

---
#### Tommy Waite solved on 5/31/2024, 03:18 AM, 0 votes: 

<p>Finally managed to get it working. It came down to:</p>
<ul>
<li>Existing installations in my Notification Hub using GCM that I had to manually delete.</li>
<li>DeviceInstallationService class to now use Microsoft.Azure.NotificationHubs.NotificationHubClient with FcmV1Installation to create/update the installation.</li>
<li>Updating the code inside the overridden OnNewToken method to use Microsoft.Azure.NotificationHubs.NotificationHubClient to update the device installation with the token.</li>
</ul>

**Comment by jkommeren** [comment]
I migrated last week. I didn't have to change anything on the client side, just like the documentation suggested.

What the documentation did not mention is that the json format of data messages is different for fcm. (Data messages always trigger the method you mentioned. In background and in foreground. Regular notifications don't iirc.) that is if you're trying to send a data message. Are you?

I found out with some trial and error. I can send you the details tomorrow if you like.

**Comment by samhouts** [comment]
@moljac

**Comment by moljac** [comment]
Thanks for the feedback.

> Since applying the changes, I can no longer get the overridden method OnMessageReceived(RemoteMessage p0) to trigger anymore, implemented from Firebase.Messaging.FirebaseMessagingService.

Can you get logcat from the device?
 
> I can see the device is registered using FCM v1 protocol, if I attempt to retrieve the installation using Microsoft.Azure.NotificationHubs.NotificationHubClient after running the code shown in the DeviceInstall.PNG image.
> 
> Wondering if there are more defined/documented steps to migrate this across or if anyone can help me get this back up and working using FCM v1?

Sorry, but no. While we (I) do the bindings, but we are not the experts in APIs of almost 700 artifacts used by Android.

**Comment by moljac** [comment]
@jkommeren Thanks for the feedback.

> I migrated last week. I didn't have to change anything on the client side, just like the documentation suggested.
> 
> What the documentation did not mention is that the json format of data messages is different for fcm. (Data messages always trigger the method you mentioned. In background and in foreground. Regular notifications don't iirc.) that is if you're trying to send a data message. Are you?

I cannot tell anything from original post.
 
> I found out with some trial and error. 

> I can send you the details tomorrow if you like.

Would you be so kind and post it here please?
